azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/features/VirtualMachineApiLiveTest.java

```diff
@@ -0,0 +1,269 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jclouds.azurecompute.arm.features;
+
+import com.google.common.base.Predicate;
+import org.jclouds.azurecompute.arm.domain.DataDisk;
+import org.jclouds.azurecompute.arm.domain.DiagnosticsProfile;
+import org.jclouds.azurecompute.arm.domain.HardwareProfile;
+import org.jclouds.azurecompute.arm.domain.IdReference;
+import org.jclouds.azurecompute.arm.domain.ImageReference;
+import org.jclouds.azurecompute.arm.domain.NetworkInterfaceCard;
+import org.jclouds.azurecompute.arm.domain.NetworkProfile;
+import org.jclouds.azurecompute.arm.domain.OSDisk;
+import org.jclouds.azurecompute.arm.domain.OSProfile;
+import org.jclouds.azurecompute.arm.domain.StorageProfile;
+import org.jclouds.azurecompute.arm.domain.StorageService;
+import org.jclouds.azurecompute.arm.domain.VHD;
+import org.jclouds.azurecompute.arm.domain.VirtualMachine;
+import org.jclouds.azurecompute.arm.domain.VirtualMachineInstance;
+import org.jclouds.azurecompute.arm.domain.VirtualMachineProperties;
+import org.jclouds.azurecompute.arm.functions.ParseJobStatus;
+import org.jclouds.azurecompute.arm.internal.BaseAzureComputeApiLiveTest;
+import org.jclouds.util.Predicates2;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.testng.Assert.assertTrue;
+
+@Test(groups = "live", testName = "VirtualMachineApiLiveTest")
+public class VirtualMachineApiLiveTest extends BaseAzureComputeApiLiveTest {
+
+   private String subscriptionid = getSubscriptionId();
+   private String vmName = null;
+   private String nicName = null;
+
+   @BeforeClass
+   public void Setup() {
+      NetworkInterfaceCard nic = getOrCreateNetworkInterfaceCard(NETWORKINTERFACECARD_NAME);
+      assertNotNull(nic);
+      nicName = nic.name();
+   }
+
+   private String getName() {
+      if (vmName == null) {
+         vmName = String.format("%3.24s",
+                 System.getProperty("user.name") + RAND + this.getClass().getSimpleName()).toLowerCase().substring(0, 15);
+      }
+      return vmName;
+   }
+
+   @Test
+   public void testCreate() {
+      StorageAccountApi storageApi = api.getStorageAccountApi(getResourceGroupName());
+      StorageService storageAccount = storageApi.get(getStorageServiceName());
+      String blob = storageAccount.storageServiceProperties().primaryEndpoints().get("blob");
+
+      String id = "/subscriptions/" + subscriptionid + "/resourceGroups/" + getResourceGroupName() +
+              "/providers/" + "Microsoft.Compute/virtualMachines/" + getName();
```

Comment:
Will users have to manually build this string every time a VM has to be created? It doesn't look very usable. I think it could be a way to get rid of this ID parameter in the method.

First, change the payload defined in the interface to something like the following, to configure the ID template, and remove the ID parameter from the method signature:

```java
@Payload("%7B\"id\":\"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroup}" 
    + "/providers/Microsoft.Compute/virtualMachines/{name}\","
    + "\"name\":\"{name}\",\"type\":\"Microsoft.Compute/virtualMachines\","
    + "\"location\":\"{location}\",\"tags\":%7B%7D,\"properties\":{properties}%7D")
```

Now we just need to resolve the `subscriptionId` and `resourceGroup` parameters and put them on the payload. The resourceGroup has been already passed to the `api.getVirtualMachineApi` method so we should have it, and the subscriptionId is part of the endpoint. We could automatically get them as follows:

In the `AzureComputeHttpApiModule`, add a method that provides the subscription id, as follows:

```java
@Provides
@Named("subscription-id")
protected String provideSubscriptionId(ProviderMetadata providerMetadata) {
  // Return the subscription id extracted from the providerMetadata.getEndpoint()
}
```

Then create a custom binder that extends the `BindToJsonPayload` like

```java
@Singleton
public class BindSubscriptionIdAndResourceName extends BindToJsonPayload {
   private final String subscriptionId;

   @Inject public BindSubscriptionIdAndResourceName(Json jsonBinder,
         @Named("subscription-id") String subscriptionId) {
      super(jsonBinder);
      this.subscriptionId = subscriptionId;
   }
   
   @Override public <R extends HttpRequest> R bindToRequest(R request, Map<String, Object> postParams) {
      checkArgument(request instanceof GeneratedHttpRequest, "this binder only supports GeneratedHttpRequests");
      checkArgument(GeneratedHttpRequest.class.cast(request).getCaller().isPresent(), "this binder is only for delegate apis");

      // Get the value for the resource group name from the caller API
      Invocation caller = GeneratedHttpRequest.class.cast(request).getCaller().get();
      String resourceGroupName = (String) caller.getArgs().get(0);

      // Populate the computed values to the payload param map so they are bound to the payload
      postParams.put("resourceGroup", resourceGroupName);
      postParams.put("subscriptionId", subscriptionId);

      return super.bindToRequest(request, postParams);
   }
}
```

Finally, change the `@MapBinder` annotation in the method to use this new binder class.

I haven't tested it, but it should work or at least give a hint on how we can infer the subscription id and resource name in the method so users don't have to generate that ID.